### PR TITLE
Allow custom workspace name

### LINKED ISSUE #73: Allow for renaming the `workspace` folder
Currently `run` assumes the workspace will be available at `<repo_path>/workspace` and this is a good default I think, but it would be nice if you could rename that.
For example, some projects require that the folder have a different name (e.g. Jekyll repositories used for GitHub.io pages require that the source documents are in a folder called docs which cannot be in a sub-directory).
In this case it would be good to be able to specify somewhere what that workspace folder is called.
The most appropriate place at the moment is the Makefile (although it may be better to remove use of a Makefile in future #60).

### PR DESCRIPTION
**Pull request by markhedleyjones**
Closes #73 

This pull request adds the ability to control the name of the workspace.
It also moves the workspace from the operating system's root directory into the user account used in the image.
In doing this, the ability to set user accounts in the image has also been added.

Also, the `.docker-bbq` file now contains variables that control how to mount the workspace and which workdir to use, and variables to disable prideledged mode and automatic workspace mountind (which are set depending on the type of build you do).

Finally, because this considerably modifies the format of the Makefile and Dockerfile, a new option has been added to `create-repo` called `--regenerate`.
When running this option from within a repository, it will recreate the Makefile and Dockerfile to match the latest template version. This can be used to upgrade an older docker-bbq repository when any incompatibilities occur.

### CODE REVIEW

**Comment by markhedleyjones** [comment]
Also, the behaviour of `run` has changed. Previously it would try to automatically determine the correct repository from a path in the first argument (so you could be outside of the workspace entirely and it would still find the correct workspace). That wasn't so reliable because if you wanted to call something like roslaunch ... then it couldn't determine where the image was. So, instead of that, a `--repo` parameter has been added so you can specify the path to the repository if you're not currently inside of it's workspace. It's not as handy as automatically determining from the arguments, but it's much more robust. That functionality was included in this pull-request because as name of the workspace changed, the previous assumption of the workspace's path always being known was no longer valid and was part of how paths were being translated between the host-os and the container.

File: bin/create-repo [hunk 1]
```diff
@@ -61,41 +67,43 @@ case $i in
...
 
-if [[ $# -eq 1 ]]; then
-  echo "Please specify a name for the repository!"
-fi
 
-repo_name="$2"
+if [[ ${regenerate_repository} == 0 ]]; then
+  if [[ $# -eq 1 ]]; then
+    echo "Please specify a name for the repository!"
+  fi
```
Comment (f-fl0) [inline bin/create-repo:87]: I think the script should exit with an error code in this case.

File: bin/create-repo [hunk 2]
```diff
@@ -115,36 +123,61 @@ compile_file() {
...
-# Generate .gitignore
-cat ${path_templates}/components/gitignore  > "${repo_name}"/.gitignore
-for component in ${gitignore_entries[*]}; do
-  echo "${component}" >> "${repo_name}"/.gitignore
-done
+  # Generate .gitignore
+  cat ${path_templates}/components/gitignore  > "${repo_name}"/.gitignore
+  for component in ${gitignore_entries[*]}; do
+    echo "${component}" >> "${repo_name}"/.gitignore
+  done
```
Comment (f-fl0) [inline bin/create-repo]: [optional] Maybe this should only be done in `enable_git` is 1.

File: tests/definitions/run-command.sh [hunk 3]
```diff
@@ -4,36 +4,40 @@ create-repo debian ${TESTREPO} > /dev/null 2>&1
...
+cd /tmp
+name="Run target outside repository"
+pass "run --repo ${repo_path} workspace/target.sh | grep boom"
 
 
 # Build production image with target script
 cd ${repo_path}
 make production > /dev/null 2>&1
-rm workspace/target.sh
+#rm workspace/target.sh
```
Comment (f-fl0) [inline tests/definitions/run-command.sh]: Is this still needed?

**Review by f-fl0 — COMMENTED** [review]
I just went through the changes and left a few comments. I didn't have the time to test yet but I will do that asap.

File: tests/definitions/run-command.sh (hunk 3 above)
Comment (markhedleyjones) [reply tests/definitions/run-command.sh]: Definitely - good spotting! The target should exist inside the production image so I wanted to remove it from the host to make sure it's not executing the host's copy.

**Review by f-fl0 — APPROVED** [review]
LGTM

**Pull request merged**